crucible/src/Lang/Crucible/Backend/Prove.hs

```diff
@@ -3,32 +3,212 @@ Module      : Lang.Crucible.Backend.Prove
 Description : Proving goals under assumptions
 Copyright   : (c) Galois, Inc 2024
 License     : BSD3
+
+== Overview
+
+The functions in this module dispatch proof obligations arising from symbolic
+execution using SMT solvers. There are several dimensions of configurability,
+encapsulated in a 'ProofStrategy':
+
+* Offline vs. online: Offline solvers ('offlineProver') are simpler to manage
+  and more easily parallelized, but starting processes adds overhead; online
+  solvers ('onlineProver') can share state as assumptions are added. See the
+  top-level README for What4 for further discussion of this choice.
+* Failing fast ('failFast') vs. keeping going ('keepGoing')
+* Timeouts: Proving with timeouts ('offlineProveWithTimeout') vs. without
+  ('offlineProve')
+
+Once an appropriate strategy has been selected, it can be passed to entrypoints
+such as 'proveObligations' to dispatch proof obligations.
+
+== Parallelism
```

Comment:
Having worked on it for a bit, I'm not sure it actually makes sense to export helpers for running proofs in parallel from this module. The upsides are:

- Parallel proving is complex, it's nice to have a single, centralized implementation.
- The approach used here will probably work for a variety of clients.

But there are some downsides as well:

- There are some knobs that are currently not exposed. This means this implementation won't work for everybody, but exposing them would make the API more complex. Examples include the delay when polling threads, and the maximum number of threads.
- The implementation is complex, so perhaps hard to understand, extend, and maintain.
- The implementation has some limitations (e.g., it doesn't support failing fast, in part because this would require a more complex interface that wouldn't match the single-threaded version).

More featureful clients might just want to roll their own custom traversal of `Goals` which incorporates whatever features they need.

When reviewing, keep in mind that I'd be aligned with simply reverting the commits having to do with parallelism.

Reply:
The following are vague generalities you might want to just ignore but might be somewhat useful for thinking about this:

- It seems that it should be possible to roll config settings (like delay values) into the API in a way that allows basic users to mostly ignore them. Haskell doesn't make this easy, but it should still be possible...
- In general providing an incomplete/special case library that some users will need to reimplement doesn't seem all that desirable: it incurs most of the costs (both immediate and long-term) without actually solving the problem for keeps. Furthermore, the reimplementations will probably end up being cut-paste-and-edit which isn't good for anyone in the long term. I think, though, this doesn't mean "don't do it", it means "go ahead and do it right".
- If doing it right involves more work than you're prepared to do right now, the right positioning is probably not "some clients will need to reimplement this" but "for some users we will need to extend this"; document the limitations and a sketch of what's entailed in lifting them.
- If none of the above is feasible it's probably also better to split the just parts that would need to be rewritten into a separate module (or even a separate library) so that it's clear where the boundaries are.

It also might make sense to split the parallel bits into a separate branch so they're more readily separable later on.

Reply:
Yep, I'm convinced! Reverted the bits about parallelism.